feat(hooks): add useCollectionQuery and useIsMounted [JOB-32207] (#561)
GetJobber/atlantis: packages/hooks/src/useCollectionQuery/useCollectionQuery.mdx

---
name: useCollectionQuery
menu: Hooks
route: /hooks/use-collection-query
---

import { Playground } from "docz";
import { List } from "@jobber/components/List";
import { Button } from "@jobber/components/Button";
import { Spinner } from "@jobber/components/Spinner";
import { Divider } from "@jobber/components/Divider";
import { useCollectionQuery } from "@jobber/hooks";
import {
  apolloClient,
  LIST_QUERY,
  mapDataToListItem,
  styles,
  propsList,
  subscriptionPropsList,
} from "./mdxUtils";

# UseCollectionQuery

> `useCollectionQuery` is a generalized hook that that fetches a collection of
> data. For use with any lists populated by graphQL collections.

Reviewer: Just thinking that it does more than just fetch. 😄 

```suggestion
`useCollectionQuery` is a generalized hook that that aids in the interaction of data collections. For use with any lists populated by graphQL collections.
```

      loadingNextPage,
      loadingInitialContent,
    } = useCollectionQuery({
      // useCollectionQuery should be called with the query type, and
      // optionally, the subscription type. The playground errors with
      // typing included, so typing has been removed in this example.
      // Please see the first example for appropriate typing.
      query: LIST_QUERY,
      queryOptions: {
        fetchPolicy: "network-only",
        nextFetchPolicy: "cache-first",
        client: apolloClient,
      },
      getCollectionByPath(items) {
        return items.allPlanets;
      },
    });
    const items = mapDataToListItem(data) || [];
    return (
      <>
        <div style={styles.spinners}>
          <div style={styles.loadingIndicator}>
            <h4>Initial Loading Indicator</h4>
            {loadingInitialContent && <Spinner />}
          </div>
          <div style={styles.loadingIndicator}>
            <h4>Refreshing Indicator</h4>
            {loadingRefresh && <Spinner />}
          </div>
          <div style={styles.loadingIndicator}>
            <h4>Fetching More Indicator</h4>
            {loadingNextPage && <Spinner />}
          </div>

>         </div>

Reviewer: It's a bit of an abuse of intention but, it might be easier to just use a `Button` component with the `disabled` prop here rather than a bunch of custom styling. Just thinking if we can strip out some of the boiler plate it can better keep the example focused on the workings of the hook itself.

@chris-at-jobber thoughts?

Author: @seanhealy This is an interesting idea. I am curious what @chris-at-jobber thinks. Would these buttons not be functional other then being disabled / enabled?

Reviewer 2: I am not sure I understand what is being suggested, or why 😅  What would the buttons replace? Why would they be disabled?

Author: @chris-at-jobber My interpretation is that there would be three buttons that did nothing when clicked. These buttons would be disabled when not in a loading state, and enabled if they were in a loading state. The code becomes much simpler, and does not need styling. I actually really like this change! I have recorded a demo for you :) 

https://user-images.githubusercontent.com/6353164/123437099-52c29e00-d59d-11eb-8773-d040ffa8aa40.mp4

Reviewer 2: Hmmm

I don't think I understand what those buttons are actually communicating, and they flicker in and out of "active" to quickly to be of any real help. I thought the previous approach of "header + spinner" was being used to visualize that different loading paths are hit in the code for different types of loading, but if that's not what we're trying to do, maybe just remove the whole section? It seems like the list content updating when you trigger the actions is good enough.

        <Divider />
        <div>

>           {items && (

Reviewer: Might be easier to use a `List` here.

https://atlantis.getjobber.com/components/list

Author: @seanhealy Sorry -- I am unclear on what you mean. I am using a List here.

Reviewer: This is just me not reading properly.

            <>

>               <h4>Star Wars Planets</h4>

Reviewer: https://atlantis.getjobber.com/packages-components-src-heading-heading

              <List items={items} />
            </>
          )}
        </div>
        <Divider />
        <div style={styles.actions}>
          <Button
            label={"Reload"}
            onClick={() => {
              refresh();
            }}
          />
          <Button
            label={"Fetch More"}
            onClick={() => {
              nextPage();
            }}
          />
        </div>
      </>
    );
  }}
</Playground>

## Props

> <Playground>
>   {() => {
>     return <List items={propsList} />;
>   }}
> </Playground>

Reviewer: You should be able to just do?

```suggestion
<List items={propsList} />
```

## Subscription props

<Playground>
  {() => {
    return <List items={subscriptionPropsList} />;
  }}
</Playground>

> 

Reviewer: I think it would also be nice to talk about `useCollectionQuery`s return value to document what they provide / do.

```
{
    data,
    error,
    refresh,
    nextPage,
    loadingRefresh,
    loadingNextPage,
    loadingInitialContent,
}
```

Author: Great idea! @seanhealy
